pkg/ddl/backfilling.go

```diff
@@ -619,6 +621,146 @@ func SetBackfillTaskChanSizeForTest(n int) {
 	backfillTaskChanSize = n
 }
 
+func (dc *ddlCtx) runAddIndexInLocalIngestMode(
+	ctx context.Context,
+	sessPool *sess.Pool,
+	t table.PhysicalTable,
+	reorgInfo *reorgInfo,
+) error {
+	// TODO(tangenta): support adjust worker count dynamically.
+	startKey, endKey := reorgInfo.StartKey, reorgInfo.EndKey
+	if err := dc.isReorgRunnable(reorgInfo.Job.ID, false); err != nil {
+		return errors.Trace(err)
+	}
+	job := reorgInfo.Job.Clone()
+	ctx = tidblogutil.WithCategory(ctx, "ddl-ingest")
+	opCtx := NewLocalOperatorCtx(ctx, job.ID)
+	bcCtx, err := getBackendCtx(ctx, dc.store, dc.etcdCli, job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	idxCnt := len(reorgInfo.elements)
+	indexIDs := make([]int64, 0, idxCnt)
+	indexInfos := make([]*model.IndexInfo, 0, idxCnt)
+	uniques := make([]bool, 0, idxCnt)
+	for _, e := range reorgInfo.elements {
+		indexIDs = append(indexIDs, e.ID)
+		indexInfo := model.FindIndexInfoByID(t.Meta().Indices, e.ID)
+		if indexInfo == nil {
+			logutil.DDLIngestLogger().Warn("index info not found",
+				zap.Int64("table ID", t.Meta().ID),
+				zap.Int64("index ID", e.ID))
+			return errors.Errorf("index info not found: %d", e.ID)
+		}
+		indexInfos = append(indexInfos, indexInfo)
+		uniques = append(uniques, indexInfo.Unique)
+	}
+
+	engines, err := bcCtx.Register(indexIDs, uniques, job.TableName)
+	if err != nil {
+		tidblogutil.Logger(opCtx).Error("cannot register new engine",
+			zap.Error(err),
+			zap.Int64s("index IDs", indexIDs))
+		return errors.Trace(err)
+	}
+	defer bcCtx.UnregisterEngines()
+	sctx, err := sessPool.Get()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer sessPool.Put(sctx)
+
+	cpMgr, err := ingest.NewCheckpointManager(
+		ctx,
+		bcCtx,
+		sessPool,
+		job.ID,
+		indexIDs,
+		bcCtx.GetLocalBackend().LocalStoreDir,
+		dc.store.(kv.StorageWithPD).GetPDClient(),
+	)
+	if err != nil {
+		logutil.DDLIngestLogger().Warn("create checkpoint manager failed", zap.Error(err))
+	} else {
+		defer cpMgr.Close()
+		cpMgr.Reset(t.GetPhysicalID(), startKey, endKey)
```

Comment:
// Reset resets the checkpoint manager between two partitions.

This is not the expected usage of `Reset`. And I see `importStarted` logic is deleted, in this PR which line can we find that the last failure is happened after all data is flushed and we only need to import after resuming?

Reply:
If all data is flushed, we will `(src *TableScanTaskSource) adjustStartKey` according to the checkpoint manager. Thus, no data will be written again.